### ISSUE TITLE
Service Mirroring Remote Gateway

### ISSUE THREAD

**Issue by zaharidichev** [linkerd/linkerd2#3992]
The current implementation of the reference gateway looks like: 

```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: nginx-configuration
  namespace: default
data:
  nginx.conf: |-
    events {
    }
    http {
      include /etc/nginx/resolvers-data/resolvers.conf;
      server {
        location / {
          proxy_pass http://$http_host;
        }
      }
    }
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  namespace: default
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      volumes:
      - name: config
        configMap:
          name: nginx-configuration
      - name: resolvers-data
        emptyDir: {}    
      initContainers:
        - name: setup-resolvers
          image: busybox:1.28
          command: ["sh", "-c", "echo resolver $(awk 'BEGIN{ORS=\" \"} $1==\"nameserver\" {print $2}' /etc/resolv.conf)\";\" > /etc/nginx/resolvers-data/resolvers.conf"]
          volumeMounts:
            - name: resolvers-data
              mountPath: /etc/nginx/resolvers-data
      containers:
        - name: nginx
          image: nginx
          ports:
            - name: http
              containerPort: 80
          volumeMounts:
          - name: config
            mountPath: /etc/nginx
          - name: resolvers-data
            mountPath: /etc/nginx/resolvers-data      
---
apiVersion: v1
kind: Service
metadata:
  name: nginx
  namespace: default
  annotations: 
    mirror.linkerd.io/gateway-identity: <IDENTITY GOES HERE>
spec:
  ports:
  - name: incoming-port
    port: 80
    protocol: TCP
  selector:
    app: nginx
  type: LoadBalancer
```

HISTORIC: 

As part of the Service mirroring architecture we need a reference implementation of a remote gateway. For the time being this should be an nginx gateway that can do a few things: 

1. It should be able to route requests following the convention: 
{svc-name}-{cluster-name}.{namespace}.svc.cluster.local:{svc_port} -> {svc-name}.{namespace}.svc.cluster.local:{svc_port}
2. We need to smohow be able to deal with fully qualified hosts. After all from the other cluster, we should be able to reach: 

- {svc-name}-{cluster-name}.{namespace}.svc.cluster.local:{svc_port}
- {svc-name}-{cluster-name}.{namespace}
- {svc-name}-{cluster-name}

How do we handle that?

**Comment by alpeb** [comment]
For handling the multiple host options we can add to the ingress an annotation like
```
nginx.ingress.kubernetes.io/server-alias: {svc-name}-{cluster-name}-{namespace},{svc-name}-{cluster-name}
```
WDYT about expanding the set of `cluster` cli subcommands from #4070 with another subcommand for triggering the mirroring of a service? It could create the ingress resource for the service, with all the annotations like the one above (and whatever else is going to be needed for [forcing nginx to route to the appropriate port according to the authority](https://github.com/linkerd/linkerd2/pull/4028#issuecomment-584889918)), plus annotating the service with the `mirror.linkerd.io` annotations.

**Comment by zaharidichev** [comment]
@alpeb I think a command like that would make much sense and is a great idea. The only thing I am hesitant about is coupling that too tightly with nginx ingress and particularily the Ingress Resource. @grampelberg WDYT?

**Comment by grampelberg** [comment]
Honestly, I don't think we should use an ingress at all. It feels like a simple nginx.conf can do everything we need.

**Comment by adleong** [comment]
I've been going back-and-forth with this.  The nice thing about nginx-ingress is that they have a fairly sophisticated nginx config that appears to have a lot of things tuned and configured outside of just the routing logic.  If we were to write our own simple nginx.conf, we would need to figure out which of those settings to include, which of the plugins we need, etc. to get to a production-ready nginx config.

I've been testing with the nginx-ingress helm chart and this ingress config:

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: cross-cluster-ingress
  namespace: default
  annotations:
    kubernetes.io/ingress.class: "nginx"
    nginx.ingress.kubernetes.io/configuration-snippet: |
      proxy_set_header l5d-dst-override $http_host;
      grpc_set_header l5d-dst-override $http_host;

spec:
  backend:
    serviceName: old-guppy-nginx-ingress-default-backend
    servicePort: 80
```

and it's working quite well so far.  The default backend is just there so that nginx has something to resolve.  When Linkerd intercepts the request, it ignores all that and just uses the l5d-dst-override header.

**Comment by adleong** [comment]
Of course, the above requires Linkerd.  

It's remarkably hard to get an nginx config that simply routes the same way as Kubernetes does.  For example, nginx doesn't use /etc/resolv.conf which means that it must be explicitly configured with a DNS server and it will not use the search domains which means that it cannot resolve Hosts which are not fully qualified.  That can probably be resolved with some header rewriting and some regexes but yeesh.  It feels like this should be easier...

**Comment by grampelberg** [comment]
I'm sensitive to figuring out how to include an nginx-ingress install in the linkerd install vs just using nginx. That said, just using nginx might definitely be more complex.

**Comment by zaharidichev** [comment]
> It's remarkably hard to get an nginx config that simply routes the same way as Kubernetes does. 
Have been digging into that this morning and I have to agree. This is a bit of a rabbit hole. @adleong provided some quite useful regexes that can be used to rewrite to the correct host. However using an ingress resource that does not require Linkerd prooves quite hard indeed. So a few things that I found out: 

1. Adding a configuraton with a config-snippet annotation just dumps the snippet in the `server` block, which does not work for us as we are using a `map` directive that needs to live in a different block (http)
2. We can use a config map to dynamically configure nginx and add an `http-snippet` key that will modify our nginx.conf to include the map directive into the http block: 

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-nginx-nginx-ingress-controller
  namespace: default
  labels:
    app.kubernetes.io/name: ingress-nginx
    app.kubernetes.io/part-of: ingress-nginx
data:
  proxy-connect-timeout: "10"
  proxy-read-timeout: "333"
  proxy-send-timeout: "120"
  http-snippet: |
    map $http_host  $local {
      ~*http(s?)://([^.]*)-remote(.*) http$1://$2$3;
      default '';
    }
    proxy_set_header l5d-dst-override  $local;
    grpc_set_header l5d-dst-override  $local;
```

3. The problem here is that although we now have this $local set to the correct service dns (assuming its fully qualified), we dont really have a way to include a `proxy_pass` in order to redirect that. We can use `location_snippet` but trying to add a proxy_pass in there errors out because there is already a proxy_pass placed there. Moreover, this approch adds this pas to all `location` contexts in the whole nginx.conf. 

So yes.. all in all I am not quite sure what to do about all this. It seems to me that  requiring existing nginx that we can dynamically reconfigure to proxy to out internal services will be very error prone. For starters, we do not know what the present config is there. Who is to say that we are not going to mess up something that was configured by the owners of this service? 

Alternatively we can roll our own custom image with a predifined config that we know works and explain that this is a reference implementation. This will allow us to extend #4070 and bring it a bit closer to what @alpeb [suggested](https://github.com/linkerd/linkerd2/pull/4070#discussion_r381403769) and have a workflow such as. 

```
#create credentials on remote cluster
linkerd cluster create-credentials -context=remote

#lnstall service mirror
linkerd -context=local install-service-mirror | kubectl -context=local apply -f -

#Extract and add credentials in local cluster 
linkerd cluster get-credentials -context=remote | kubectl -context=local apply -f -

#Deploy gateway on remote cluster
linkerd cluster provision-gateway | kubectl -context=remote apply -f -

#Enable a serviece to be mirrored
linkerd -context=remote cluster mirror-service --gatewayName=someName --gatewayNs=someNs | kubectl -context=remote apply -f -
```


Just my two cents. Dont get me wrong, I am not pushing for providing our own nginx.conf. After playing with that a bit, I am certainly not looking forward to doing that. I simply do not see any other way that will remove all the UX friction. 

@grampelberg WDYT?

**Comment by grampelberg** [comment]
> Alternatively we can roll our own custom image with a predifined config that we know works and explain that this is a reference implementation.

This is my preference. We're going to rely on linkerd's client side and should be able to make some big shortcuts in the nginx.conf because of this. FWIW, I'm not set on using nginx either. We could do this with a little bit of envoy config as well. Don't need an xDS endpoint to do this kind of rewriting.

**Comment by zaharidichev** [comment]
@grampelberg @adleong 
Here is a nginx config that works.  This assumes tht we have the namespace in the service name. We can change that and not support routing if there is no (at least) namespace in the path (i.e. `service-ns-remote.ns`)
```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: nginx-configuration
  namespace: default
data:
  nginx.conf: |-
    events {
    }
    http {

      map $http_host $local {
        # service-ns-remote.ns:7000 => service.ns.svc.cluter.local:7000
        ~*([^.]*)-([^.]*)-[^:]*(:\d+)? $1.$2.svc.cluster.local$3;
      }

      include /etc/nginx/resolvers-data/resolvers.conf;
      server {
        location / {
          proxy_pass http://$local;
        }
      }
    }
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  namespace: default
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      volumes:
      - name: config
        configMap:
          name: nginx-configuration
      - name: resolvers-data
        emptyDir: {}    
      initContainers:
        - name: setup-resolvers
          image: busybox:1.28
          command: ["sh", "-c", "echo resolver $(awk 'BEGIN{ORS=\" \"} $1==\"nameserver\" {print $2}' /etc/resolv.conf)\";\" > /etc/nginx/resolvers-data/resolvers.conf"]
          volumeMounts:
            - name: resolvers-data
              mountPath: /etc/nginx/resolvers-data
      containers:
        - name: nginx
          image: nginx
          ports:
            - name: http
              containerPort: 80
          volumeMounts:
          - name: config
            mountPath: /etc/nginx
          - name: resolvers-data
            mountPath: /etc/nginx/resolvers-data      
---
apiVersion: v1
kind: Service
metadata:
  name: nginx
  namespace: default
spec:
  ports:
  - name: ingress
    port: 80
    protocol: TCP
  selector:
    app: nginx
  type: LoadBalancer
  ```

**Comment by grampelberg** [comment]
Woah, nice!

**Comment by grampelberg** [comment]
@zaharidichev this is done, right?

**Comment by zaharidichev** [comment]
@grampelberg yes

**Comment by zaharidichev** [comment]
This was closed by #4265
